packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/AmplifyApiPlugin.kt

```diff
@@ -49,26 +52,98 @@ class AmplifyApiPlugin : FlutterPlugin, ActivityAware, MethodCallHandler {
     when (call.method) {
       "query" ->
         onQuery(result, call.arguments as Map<String, Any>)
+      "mutate" ->
+        onMutate(result, call.arguments as Map<String, Any>)
       else -> result.notImplemented()
     }
   }
 
   fun onQuery(flutterResult: Result, request: Map<String, Any>) {
+    var document: String
+    var variables: Map<String, Any>
+
     try {
-    Log.i("AmplifyFlutter", "API.query request :$request");
-    var result: Map<String, Any> = mapOf(
-            "getTodo" to mapOf(
-                    "id" to "123",
-                    "name" to "test todo"
-            )
+      document = request["document"] as String
+      variables = request["variables"] as Map<String, Any>
+    } catch (e: ClassCastException) {
+      postFlutterError(
+              flutterResult,
+              FlutterApiFailureMessage.ERROR_CASTING_INPUT_IN_PLATFORM_CODE.toString(),
+              e)
+      return
+    } catch (e: Exception) {
+      postFlutterError(
+              flutterResult,
+              FlutterApiFailureMessage.AMPLIFY_REQUEST_MALFORMED.toString(),
+              e)
+      return
+    }
+
+    Amplify.API.query(
+        SimpleGraphQLRequest<String>(
+                document,
+                variables,
+                String::class.java,
+                GsonVariablesSerializer()
+        ),
+        {
+          var result: Map<String, Any> = mapOf(
+                  "data" to it.data,
+                  "errors" to it.errors
+          )
+          handler.post { flutterResult.success(result) }
+        },
+        {
+          postFlutterError(
+                  flutterResult,
+                  FlutterApiFailureMessage.AMPLIFY_API_QUERY_FAILED.toString(),
+                  it)
+        }
     )
-      handler.post { flutterResult.success(result) }
-  } catch (error: Exception) {
+  }
+
+  fun onMutate(flutterResult: Result, request: Map<String, Any>) {
+    var document: String
+    var variables: Map<String, Any>
+
+    try {
+      document = request["document"] as String
+      variables = request["variables"] as Map<String, Any>
+    } catch (e: ClassCastException) {
+      postFlutterError(
+              flutterResult,
+              FlutterApiFailureMessage.ERROR_CASTING_INPUT_IN_PLATFORM_CODE.toString(),
+              e)
+      return
+    } catch (e: Exception) {
       postFlutterError(
               flutterResult,
-              "ERROR_IN_PLATFORM_CODE",
-              error)
+              FlutterApiFailureMessage.AMPLIFY_REQUEST_MALFORMED.toString(),
+              e)
+      return
     }
```

Comment:
I believe this code is common between the APIs and we can extract it to a function. This can then be leveraged  by `query`, `mutate` and eventually `subscribe`.

Reply:
Given that we're setting variables that are used elsewhere in the method I don't know if it makes sense as a separate function - you'd have to return two variables which would properly be returned as a custom object so I think it's just overkill.

Reply:
I was actually thinking something more like `getDocument()`, `getVariables()`, and being able to give more specific error messages as a result. But we can address this as a follow up.